### PR TITLE
`as_strided` : add size and stride length check

### LINKED ISSUE #39281: Seg fault using `as_strided`
## 🐛 Bug

The following code seg faults:
``` python
import torch
torch.tensor(4.).as_strided((1, ), ())
# Segmentation fault: 11
```

## To Reproduce

See above.

## Expected behavior

Expected an error message.  Not a segmentation fault (which also kills the python interpreter).

## Environment

Able to reproduce on colab and on Mac OSX.  The following env is from colab.  The Mac OSX env is also with torch 1.5.0, but without cuda.

```
Collecting environment information...
PyTorch version: 1.5.0+cu101
Is debug build: No
CUDA used to build PyTorch: 10.1

OS: Ubuntu 18.04.3 LTS
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
CMake version: version 3.12.0

Python version: 3.6
Is CUDA available: No
CUDA runtime version: 10.1.243
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5

Versions of relevant libraries:
[pip3] numpy==1.18.4
[pip3] torch==1.5.0+cu101
[pip3] torchsummary==1.5.1
[pip3] torchtext==0.3.1
[pip3] torchvision==0.6.0+cu101
[conda] Could not collect
```

cc @ezyang @gchanan @zou3519

### PR DESCRIPTION
**Pull request by kshitij12345**
Fixes #39281

### CODE REVIEW

File: test/test_torch.py [hunk 1]
```diff
@@ -10898,6 +10898,11 @@ def test_empty_strided(self, device):
                 self.assertEqual(empty_strided.shape, as_strided.shape)
                 self.assertEqual(empty_strided.stride(), as_strided.stride())
 
+    def test_strided_mismatched_stride_shape(self, device):
+        for shape, strides in [((1, ), ()), ((1, 2), (1, ))]:
+            with self.assertRaisesRegex(RuntimeError, "mismatch in length of strides and shape"):
+                torch.tensor(0.42, device=device).as_strided(shape, strides)
+
```
Comment (kshitij12345) [inline test/test_torch.py:10905]: @ailzhang 

`RuntimeError` is not thrown with XLA.
Should we have something similar to what `test_vander` used to have,
```python
if self.device_type != 'XLA':
    # Test Code
```

**Review by ailzhang — APPROVED** [review]
Stamping as the test is skipped on XLA job.

File: aten/src/ATen/native/TensorShape.cpp [hunk 2]
```diff
@@ -627,6 +627,8 @@ Tensor make_qtensor(const Tensor& self, IntArrayRef size, IntArrayRef stride, Qu
 }
 
 Tensor as_strided_tensorimpl(const Tensor& self, IntArrayRef size, IntArrayRef stride, optional<int64_t> storage_offset_) {
+  TORCH_CHECK(
+      size.size() == stride.size(), "mismatch in length of strides and shape");
```
Comment (peterbell10) [inline aten/src/ATen/native/TensorShape.cpp:631]: You are missing `as_strided_` below.

**Comment by gchanan** [comment]
In light of @peterbell10's suggestion, I think we should just move the check to `setStrided` (which all these functions end up calling).

**Comment by gchanan** [comment]
And look at that, it's already checked https://github.com/pytorch/pytorch/blob/6d3e4aa0f92da0425ccc7d26908226338a9c7e5c/aten/src/ATen/native/Resize.h#L146, but should be a `TORCH_CHECK` instead of `AT_ASSERT`.

@kshitij12345 -- can you make that change and also add a test for as_strided_?  We'd like to get this into 1.5.1, so I'll ut up a PR tomorrow if you aren't able to.

File: aten/src/ATen/native/Resize.h [hunk 3]
```diff
@@ -134,6 +134,7 @@ inline void setStrided(
     IntArrayRef size,
     IntArrayRef stride,
     int64_t storage_offset) {
+  TORCH_CHECK(size.size() == stride.size(), "mismatch in length of strides and shape");
```
Comment (kshitij12345) [inline aten/src/ATen/native/Resize.h:137]: Thanks @gchanan @peterbell10 .

Have addressed the comments.

Also note that have moved the `TORCH_CHECK` at the beginning of `setStrided` as it was seg faulting if it was at the current location of `AT_ASSERT`. (Seg Fault was probably in `checkInBoundsForStorage` but didn't verify.)

**Review by gchanan — APPROVED** [review]
looks good, thanks!

**Pull request closed without merging**